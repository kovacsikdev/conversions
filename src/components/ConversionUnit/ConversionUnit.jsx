import React, { useState } from 'react';
import { formulas } from '../../libs/formulas/formulas';
import './ConversionUnit.scss';

export const ConversionUnit = ({imperial, metric}) => {
  const [imperialNum, setImperialNum] = useState(imperial === 'Fahrenheit' ? 32 : 0);
  const [metricNum, setMetricNum] = useState(0);

  const updateMetricNum = num => {
    setImperialNum(num);
    setMetricNum(formulas[imperial.toLowerCase()](num).toFixed(2));
  }

  const updateImperialNum = num => {
    setMetricNum(num);
    setImperialNum(formulas[metric.toLowerCase()](num).toFixed(2));
  }

  return (
    <div className="ConversionUnit">
      <div className="imperial">
        <label>{imperial}
          <input
            type="number"
            value={imperialNum}
            onChange={e =>{
              updateMetricNum(e.target.value)
            }}
            onFocus={e => {e.target.select();}}
          />
        </label>
      </div>
      <div className="spacer">=</div>
      <div className="metric">
        <label>{metric}
          <input
            type="number"
            value={metricNum}
            onChange={e =>{
              updateImperialNum(e.target.value)
            }}
            onFocus={e => {e.target.select();}}
          />
        </label>
      </div>
    </div>
  )
};