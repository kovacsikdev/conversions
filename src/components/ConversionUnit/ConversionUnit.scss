.ConversionUnit {
  display: flex;
  justify-content: space-between;
  padding: 1.5rem 0.75rem;

  input {
    box-sizing: border-box;
    width: 100%;
    font-size: 1.25rem;
    padding: 0.25rem 0;
    border: none;
    border-bottom: 1px solid #ccc;
    background: var(--background);
    color: var(--color);
  }

  div {
    flex: 1;
  }

  .spacer {
    max-width: 50px;
    display: flex;
    justify-content: center;
  }

  .imperial {
    text-align: right;

    input {
      text-align: right;
    }
  }
}
