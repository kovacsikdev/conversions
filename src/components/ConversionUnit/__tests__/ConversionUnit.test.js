import React from 'react';
import { shallow } from 'enzyme';
import { ConversionUnit } from '../ConversionUnit';

const mockData = {
  imperial: 'Pounds',
  metric: 'Kilograms'
}
const mockEvent = { target: { value: 10 } };
const mockSelect = { 
  target: {
    select: jest.fn()
  }
}

describe("components/ConversionUnit", () => {

  it('should render the component correctly', () => {
    const wrapper = shallow(<ConversionUnit {...mockData}/>);
    const imperialNum = wrapper.find('input').at(0).prop('value');
    const metricNum = wrapper.find('input').at(1).prop('value');

    expect(imperialNum).toEqual(0);
    expect(metricNum).toEqual(0);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render the Fahrenheit unit with a default of 32', () => {
    const wrapper = shallow(
      <ConversionUnit
        imperial="Fahrenheit"
        metric="Celsius"
      />
    );
    const fahrenheitNum = wrapper.find('input').at(0).prop('value');

    expect(fahrenheitNum).toBe(32);
  });

  it('should update the metric unit if the imperial unit is changed', () => {
    const wrapper = shallow(<ConversionUnit {...mockData}/>);
    const imperialInput = wrapper.find('input').at(0);

    // Update input of imperial unit
    imperialInput.simulate('change', mockEvent);

    // Get value of metric input
    const metricNum = wrapper.find('input').at(1).prop('value');

    // Metric input should not be zero anymore
    expect(metricNum).not.toEqual(0);
  });

  it('should update the imperial unit if the metric unit is changed', () => {
    const wrapper = shallow(<ConversionUnit {...mockData}/>);
    const metricInput = wrapper.find('input').at(1);

    // Update input of metric unit
    metricInput.simulate('change', mockEvent);

    // Get value of imperial unit
    const imperialNum = wrapper.find('input').at(0).prop('value');

    // Imperial input should not be zero anymore
    expect(imperialNum).not.toEqual(0);
  });

  it('should automatically select the input fields on focus', () => {
    // Need to rethink this test and figure out how to check if the input field
    // is automatically selected on focus
    const wrapper = shallow(<ConversionUnit {...mockData}/>);
    const imperialInput = wrapper.find('input').at(0);
    const metricInput = wrapper.find('input').at(1);

    metricInput.simulate('focus', mockSelect);
    expect(mockSelect.target.select).toHaveBeenCalledTimes(1);

    imperialInput.simulate('focus', mockSelect);
    expect(mockSelect.target.select).toHaveBeenCalledTimes(2);
  });
});