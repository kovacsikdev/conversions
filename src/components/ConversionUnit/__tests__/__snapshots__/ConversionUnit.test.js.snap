// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ConversionUnit should render the component correctly 1`] = `
<div
  className="ConversionUnit"
>
  <div
    className="imperial"
  >
    <label>
      Pounds
      <input
        onChange={[Function]}
        onFocus={[Function]}
        type="number"
        value={0}
      />
    </label>
  </div>
  <div
    className="spacer"
  >
    =
  </div>
  <div
    className="metric"
  >
    <label>
      Kilograms
      <input
        onChange={[Function]}
        onFocus={[Function]}
        type="number"
        value={0}
      />
    </label>
  </div>
</div>
`;
